### DISCUSSION TITLE
[11 Settembre 2020] Modifica CPU

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#65] [Modifiche architettura MIPS (M-Z)]
![Esame](https://github.com/Elia-Belli/ae2223/assets/107944476/1761cd9d-01b0-4049-8376-fbaf3d364c94)
![CPU](https://github.com/Elia-Belli/ae2223/assets/107944476/0add97c3-e1ca-4e88-ac83-026b17a79460)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: 
- [Soluzione.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665845/AE.-.Appello.2020-09-11.-.Esercizio.modifica.CPU.soluzione.pdf)
- [Disegno v1.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665853/AE.-.Appello.2020-09-11.-.Esercizio.modifica.CPU.soluzione.pdf)
- [Disegno v2.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665858/AE.-.Appello.2020-09-11.-.Esercizio.modifica.CPU.-.CPU.-.v2.pdf)

**Comment by FeddyLix17** [comment]
## 1) Modifiche all'architettura

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/1b9a6989-999a-41e4-800a-fc0aa9fc3ab7)


## 2) Word che esprime l'istruzione *diff_and_store $10, $11, 64*

- l'Opcode dell'istruzione sum_ w_ mem (fornito in esadecimale dalla traccia) è uguale a $\Large 0x38_{16}  =  00111000_2$ , codificato in 6 bit 111000

- il campo $rs, rappresentate il primo registro sorgente, è il registro numero 10 ($t2) , codificato in 5 bit come 01010

- il campo $rt, rappresentante il secondo registro sorgente, è il registro numero 11 ($t3) , codificato in 5 bit come 01011

- il campo immediate è uguale a $\Large 64_{10}$, codificato in 16 bit come  0000000001000000

![image](https://github.com/sapienzastudentsnetwork/archelab2223/assets/66442720/1b2c6abc-d2d3-4ab0-85b9-05476ff4f665)

## 3) Durata totale ciclo di clock

Considerando

- la fase di *Instruction Fetch*, nella quale si accederà alla memoria, con tempo *200 ns*

- la fase di *Instruction Decode*, nella quale si accederà ai registri, con tempo *50 ns*

- la fase di *Execute*, nella quale verrà calcolata la differenza tra $rs e $rt , con tempo *100 ns* 

- una fase nella quale verrà calcolato il valore assoluto della differenza tra $rs e $rt, con tempo *100 ns*

- la fase di *Memory access*, nella quale si accederà nuovamente alla memoria, con tempo *200 ns*

- la fase di *Write Back*, nella quale si accederà nuovamente ai registri per andare a conservare in *addr* il risultato della precedente differenza, con tempo *50 ns*

la durata totale di un ciclo di clock tale da permettere l'esecuzione della nuova istruzione sarà uguale a

$$
    \Large 50 * 2 + 100 * 2 + 200 * 2 = 700 ns
$$

**Reply by matypist** [reply]
Il disegno (svolgimento del **punto n.1**) della proposta di soluzione di @FeddyLix17 coincide con il [Disegno v1.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665853/AE.-.Appello.2020-09-11.-.Esercizio.modifica.CPU.soluzione.pdf) proposto nella soluzione del professore; il calcolo della durata totale del ciclo di clock (svolgimento del **punto n. 3**) probabilmente include un Write Back aggiuntivo rispetto al necessario, visto che l'esercizio non prevede la scrittura in un registro.

**Comment by CiottoloMaggico** [comment]
![immagine](https://github.com/sapienzastudentsnetwork/archelab2223/assets/79638284/7c5ce6d0-7834-4e34-8f2d-186d20e716f7)

**Reply by matypist** [reply]
Il disegno (svolgimento del **punto n.1**) della proposta di soluzione di CiottoloMaggico coincide con il [Disegno v1.pdf](https://github.com/sapienzastudentsnetwork/archelab2223/files/11665853/AE.-.Appello.2020-09-11.-.Esercizio.modifica.CPU.soluzione.pdf) proposto nella soluzione del professore; anche il calcolo della durata totale del ciclo di clock (svolgimento del **punto n. 3**) mi sembra corretto :ballot_box_with_check:
